microsoft/testsuites/nvme/performance.py

```diff
@@ -0,0 +1,99 @@
+
+from assertpy import assert_that
+
+from lisa import (
+    Node,
+    TestCaseMetadata,
+    TestSuite,
+    TestSuiteMetadata,
+    notifier,
+    simple_requirement,
+)
+from lisa.environment import Environment
+from lisa.features import Nvme, NvmeSettings
+from lisa.notifier import FIOResult, FIOResultsMessage
+from lisa.tools import Echo, Fio, Lscpu
+
+FIOMODES = Enum(
+    "FIOMODES",
+    [
+        "randread",
+        "randwrite",
+        "read",
+        "write",
+    ],
+)
+
+
+@TestSuiteMetadata(
+    area="nvme",
+    category="performance",
+    description="""
+    This test suite is to validate NVMe disk performance of Linux VM using fio tool.
+    """,
+)
+class nvmeperf(TestSuite):  # noqa
+    TIME_OUT = 3000
+
+    @TestCaseMetadata(
+        description="""
+        This test case uses fio to test NVMe disk performance.
+        """,
+        priority=2,
+        timeout=TIME_OUT,
+        requirement=simple_requirement(
+            supported_features=[NvmeSettings(disk_count=8)],
+        ),
+    )
+    def verify_nvme_performance(self, node: Node, environment: Environment) -> None:
+        nvme = node.features[Nvme]
+        nvme_namespaces = nvme.get_namespaces()
+        assert_that(len(nvme_namespaces)).described_as(
+            "At least 1 NVMe disk for fio testing."
+        ).is_greater_than(0)
+        filename = ":".join(nvme_namespaces)
+        echo = node.tools[Echo]
+        # This will have kernel avoid sending IPI to finish I/O on the issuing CPUs
+        # if they are not on the same NUMA node of completion CPU.
+        # This setting will give a better and more stable IOPS.
+        for nvme_namespace in nvme_namespaces:
+            # /dev/nvme0n1 => nvme0n1
+            disk_name = nvme_namespace.split("/")[-1]
+            echo.write_to_file(
+                "0",
+                node.get_pure_path(f"/sys/block/{disk_name}/queue/rq_affinity"),
+                sudo=True,
+            )
+        cpu = node.tools[Lscpu]
+        start_qdepth = cpu.get_core_count()
+        max_qdepth = start_qdepth * 256
+        numjob = cpu.get_core_count()
+        numjobiterator = 0
+        fio_result_list: List[FIOResult] = []
+        fio = node.tools[Fio]
+        for mode in FIOMODES:
+            qdepth = start_qdepth
+            while qdepth <= max_qdepth:
+                iodepth = int(qdepth / numjob)
+                fio_result = fio.launch(
+                    name=f"iteration{numjobiterator}",
+                    filename=filename,
+                    mode=mode.name,
+                    iodepth=int(iodepth),
+                    numjob=numjob,
+                )
+                fio_result_list.append(fio_result)
+                qdepth = qdepth * 2
+                numjobiterator += 1
+        fio_message = FIOResultsMessage()
+        fio_message.test_case_name = "verify_nvme_performance"
+        fio_message.block_size = 4
+        fio_message.disk_setup_type = "8 NVMe Disks"
```

Comment:
What's other values of the setup type? If all of them are disk count, we can rename it to "disk_count". The structured data is easier to compare.

Reply:
others like RAID 0 - 24 SSD Disks.

Reply:
we can add another field called disk_count, this field just put "RAID 0". This field presents how the disk is setup, not include the count of disk.

Reply:
ok, can we get the disk_count in requirement during runtime?

Reply:
It can be gotten from the node.capability.

Reply:
But it may include other image specified disks. Because this test case asked for 8 disks already. We can trust this requirement, and specify the disk count in message.